File: apps/gui/src-tauri/src/storage.rs
// Native Rust reimplementation of the TypeScript node's ShardStorage
// (apps/node/src/storage.ts). The on-disk layout and the shard-index.json
// format match byte-for-byte so shards written here are interchangeable with
// the TS daemon and vice-versa.
//
// Layout:
//   <root>/shards/<fileId[0..2]>/<fileId>_<shardIndex>.bin
//   <root>/shard-index.json  -> { version, usedBytes, shardCount, sizes }
//
// usedBytes / shardCount are O(1) reads off the in-memory index, which is
// persisted atomically (write to .tmp then rename). Missing/corrupt index is
// rebuilt by walking the shards directory.

use serde::{Deserialize, Serialize};
use sha2::{Digest, Sha256};
use std::collections::BTreeMap;
use std::io::ErrorKind;
use std::path::{Path, PathBuf};

const INDEX_VERSION: u64 = 1;

#[derive(Clone)]
struct ShardIndex {
    used_bytes: u64,
    shard_count: u64,
    // per-shard sizes keyed by `${fileId}_${shardIndex}`
    sizes: BTreeMap<String, u64>,
}

impl Default for ShardIndex {
    fn default() -> Self {
        Self {
            used_bytes: 0,
            shard_count: 0,
            sizes: BTreeMap::new(),
        }
    }
}

/// Raw shape parsed from / written to shard-index.json.
#[derive(Serialize, Deserialize)]
struct IndexFile {
    #[serde(default)]
    version: u64,
    #[serde(rename = "usedBytes")]
    used_bytes: u64,
    #[serde(rename = "shardCount")]
    shard_count: u64,
    sizes: BTreeMap<String, u64>,
}

/// Shard storage backed by the on-disk layout described above.
///
/// All public methods perform blocking filesystem I/O and are intended to be
/// called from a blocking context (e.g. `tokio::task::spawn_blocking`). The
/// type is cheap to clone via `Arc` and guards its mutable index with a plain
/// `std::sync::Mutex` (never held across an await).
pub struct ShardStorage {
    root: PathBuf,
    capacity_bytes: u64,
    index: std::sync::Mutex<ShardIndex>,
}

impl ShardStorage {
    pub fn new(root: PathBuf, capacity_bytes: u64) -> Self {
        Self {
            root,
            capacity_bytes,
            index: std::sync::Mutex::new(ShardIndex::default()),
        }
    }

    pub fn shards_dir(&self) -> PathBuf {
        self.root.join("shards")
    }

    fn index_path(&self) -> PathBuf {
        self.root.join("shard-index.json")
    }

    fn shard_path(&self, file_id: &str, shard_index: u32) -> PathBuf {
        let prefix: String = file_id.chars().take(2).collect();
        self.shards_dir()
            .join(prefix)
            .join(format!("{file_id}_{shard_index}.bin"))
    }

    fn key(file_id: &str, shard_index: u32) -> String {
        format!("{file_id}_{shard_index}")
    }

    /// Create the shards dir and load (or rebuild) the index.
    pub fn init(&self) -> std::io::Result<()> {
        std::fs::create_dir_all(self.shards_dir())?;
        self.load_index();
        Ok(())
    }

    /// Load the persisted index and trust it. If missing or corrupt, rebuild
    /// by walking the shards directory.
    fn load_index(&self) {
        if let Ok(raw) = std::fs::read_to_string(self.index_path()) {
            if let Ok(parsed) = serde_json::from_str::<IndexFile>(&raw) {
                let mut guard = self.index.lock().unwrap();
                *guard = ShardIndex {
                    used_bytes: parsed.used_bytes,
                    shard_count: parsed.shard_count,
                    sizes: parsed.sizes,
                };
                return;
            }
        }
        // missing or corrupt -> rebuild
        self.rebuild_index();
    }

    /// Walk the shards directory once and rebuild the in-memory index, then
    /// persist it.
    fn rebuild_index(&self) {
        let mut rebuilt = ShardIndex::default();
        walk_dir(&self.shards_dir(), &mut |path, size| {
            let name = path
                .file_name()
                .and_then(|n| n.to_str())
                .unwrap_or_default();
            let key = name.strip_suffix(".bin").unwrap_or(name).to_string();
            rebuilt.sizes.insert(key, size);
            rebuilt.used_bytes += size;
            rebuilt.shard_count += 1;
        });
        {
            let mut guard = self.index.lock().unwrap();
            *guard = rebuilt;
        }
        self.persist_index();
    }

    /// Persist the index atomically (write to a temp file then rename).
    fn persist_index(&self) {
        let payload = {
            let guard = self.index.lock().unwrap();
            IndexFile {
                version: INDEX_VERSION,
                used_bytes: guard.used_bytes,
                shard_count: guard.shard_count,
                sizes: guard.sizes.clone(),
            }
        };
        if let Ok(json) = serde_json::to_string(&payload) {
            let index_path = self.index_path();
            let tmp = self.root.join("shard-index.json.tmp");
            if std::fs::write(&tmp, json).is_ok() {
                let _ = std::fs::rename(&tmp, &index_path);
            }
        }
    }

    /// Store a shard, verifying its SHA-256 against `expected_hash` before
    /// writing and enforcing the capacity limit on the net delta.
    pub fn store(
        &self,
        file_id: &str,
        shard_index: u32,
        data: &[u8],
        expected_hash: &str,
    ) -> Result<u64, String> {
        let actual_hash = hex::encode(Sha256::digest(data));
        if actual_hash != expected_hash {
            return Err(format!(
                "Shard hash mismatch: got {}..., expected {}...",
                &actual_hash[..actual_hash.len().min(16)],
                &expected_hash[..expected_hash.len().min(16)]
            ));
        }

        let key = Self::key(file_id, shard_index);
        let incoming_size = data.len() as u64;

        // Compute net delta and check capacity while holding the lock briefly.
        {
            let guard = self.index.lock().unwrap();
            let existing = guard.sizes.get(&key).copied();
            let new_used = match existing {
                None => guard.used_bytes + incoming_size,
                Some(prev) => guard.used_bytes + incoming_size - prev,
            };
            if new_used > self.capacity_bytes {
                return Err(format!(
                    "capacity exceeded: {} > {}",
                    new_used, self.capacity_bytes
                ));
            }
        }

        let path = self.shard_path(file_id, shard_index);
        let prefix: String = file_id.chars().take(2).collect();
        std::fs::create_dir_all(self.shards_dir().join(prefix))
            .map_err(|e| format!("mkdir failed: {e}"))?;
        std::fs::write(&path, data).map_err(|e| format!("write failed: {e}"))?;

        // Update counters after a successful write.
        {
            let mut guard = self.index.lock().unwrap();
            let existing = guard.sizes.get(&key).copied();
            match existing {
                None => {
                    guard.shard_count += 1;
                    guard.used_bytes += incoming_size;
                }
                Some(prev) => {
                    guard.used_bytes = guard.used_bytes + incoming_size - prev;
                }
            }
            guard.sizes.insert(key, incoming_size);
        }
        self.persist_index();
        Ok(incoming_size)
    }

    /// Read a shard from disk. Returns `None` if it does not exist.
    pub fn retrieve(&self, file_id: &str, shard_index: u32) -> std::io::Result<Option<Vec<u8>>> {
        match std::fs::read(self.shard_path(file_id, shard_index)) {
            Ok(buf) => Ok(Some(buf)),
            Err(e) if e.kind() == ErrorKind::NotFound => Ok(None),
            Err(e) => Err(e),
        }
    }

    /// Remove a shard. Returns true if a file was actually unlinked. Keeps the
    /// index consistent even when the file is already gone.
    pub fn remove(&self, file_id: &str, shard_index: u32) -> std::io::Result<bool> {
        let key = Self::key(file_id, shard_index);
        match std::fs::remove_file(self.shard_path(file_id, shard_index)) {
            Ok(()) => {
                {
                    let mut guard = self.index.lock().unwrap();
                    if let Some(removed) = guard.sizes.remove(&key) {
                        guard.used_bytes = guard.used_bytes.saturating_sub(removed);
                        guard.shard_count = guard.shard_count.saturating_sub(1);
                    }
                }
                self.persist_index();
                Ok(true)
            }
            Err(e) if e.kind() == ErrorKind::NotFound => {
                // File already gone; reconcile the index if it still tracks it.
                let reconciled = {
                    let mut guard = self.index.lock().unwrap();
                    if let Some(removed) = guard.sizes.remove(&key) {
                        guard.used_bytes = guard.used_bytes.saturating_sub(removed);
                        guard.shard_count = guard.shard_count.saturating_sub(1);
                        true
                    } else {
                        false
                    }
                };
                if reconciled {
                    self.persist_index();
                }
                Ok(false)
            }
            Err(e) => Err(e),
        }
    }

    /// O(1) read of the tracked used bytes.
    pub fn used_bytes(&self) -> u64 {
        self.index.lock().unwrap().used_bytes
    }

    /// O(1) read of the tracked shard count.
    pub fn shard_count(&self) -> u64 {
        self.index.lock().unwrap().shard_count
    }
}

/// Recursively walk `dir`, invoking `on_file(path, size)` for every regular
/// file. Missing directories are treated as empty.
fn walk_dir(dir: &Path, on_file: &mut dyn FnMut(&Path, u64)) {
    let entries = match std::fs::read_dir(dir) {
        Ok(e) => e,
        Err(_) => return,
    };
    for entry in entries.flatten() {
        let path = entry.path();
        match entry.file_type() {
            Ok(ft) if ft.is_dir() => walk_dir(&path, on_file),
            Ok(ft) if ft.is_file() => {
                if let Ok(meta) = entry.metadata() {
                    on_file(&path, meta.len());
                }
            }
            _ => {}
        }
    }
}

File: apps/node/src/storage.ts
import { promises as fs } from 'fs';
import { join } from 'path';
import { createHash } from 'crypto';

interface ShardIndex {
  usedBytes: number;
  shardCount: number;
  sizes: Record<string, number>; // keyed by `${fileId}_${shardIndex}`
}

const INDEX_VERSION = 1;

export class ShardStorage {
  private readonly root: string;
  private readonly capacityBytes: number;

  private index: ShardIndex = { usedBytes: 0, shardCount: 0, sizes: {} };
  private indexLoaded = false;

  // Serialize index persistence so concurrent writes never corrupt the file.
  private persistChain: Promise<void> = Promise.resolve();
  private persistQueued = false;

  constructor(root: string, capacityBytes: number = Infinity) {
    this.root = root;
    this.capacityBytes = capacityBytes;
  }

  async init() {
    await fs.mkdir(this.shardsDir(), { recursive: true });
    await this.loadIndex();
  }

  private shardsDir() {
    return join(this.root, 'shards');
  }

  private indexPath() {
    return join(this.root, 'shard-index.json');
  }

  private shardPath(fileId: string, shardIndex: number) {
    const prefix = fileId.slice(0, 2);
    return join(this.shardsDir(), prefix, `${fileId}_${shardIndex}.bin`);
  }

  private key(fileId: string, shardIndex: number) {
    return `${fileId}_${shardIndex}`;
  }

  /**
   * Load the persisted index and trust it. If it is missing or corrupt, fall
   * back to a full rebuild by walking the shards directory.
   */
  private async loadIndex() {
    if (this.indexLoaded) return;
    try {
      const raw = await fs.readFile(this.indexPath(), 'utf8');
      const parsed = JSON.parse(raw);
      if (
        parsed &&
        typeof parsed.usedBytes === 'number' &&
        typeof parsed.shardCount === 'number' &&
        parsed.sizes &&
        typeof parsed.sizes === 'object'
      ) {
        this.index = {
          usedBytes: parsed.usedBytes,
          shardCount: parsed.shardCount,
          sizes: parsed.sizes,
        };
        this.indexLoaded = true;
        return;
      }
      await this.rebuildIndex();
    } catch {
      await this.rebuildIndex();
    }
    this.indexLoaded = true;
  }

  /** Walk the shards directory once and rebuild the in-memory index. */
  private async rebuildIndex() {
    const index: ShardIndex = { usedBytes: 0, shardCount: 0, sizes: {} };
    await walkDir(this.shardsDir(), async (path) => {
      const stat = await fs.stat(path);
      const name = path.slice(path.lastIndexOf('/') + 1);
      const key = name.endsWith('.bin') ? name.slice(0, -4) : name;
      index.sizes[key] = stat.size;
      index.usedBytes += stat.size;
      index.shardCount++;
    });
    this.index = index;
    this.indexLoaded = true;
    await this.persistIndex();
  }

  /**
   * Persist the index atomically (write to a temp file then rename). Writes are
   * serialized so concurrent mutations cannot interleave and corrupt the file.
   * Multiple queued requests collapse into a single trailing write.
   */
  private persistIndex(): Promise<void> {
    if (this.persistQueued) return this.persistChain;
    this.persistQueued = true;
    this.persistChain = this.persistChain
      .catch(() => {})
      .then(async () => {
        this.persistQueued = false;
        const tmp = `${this.indexPath()}.tmp`;
        const payload = JSON.stringify({
          version: INDEX_VERSION,
          usedBytes: this.index.usedBytes,
          shardCount: this.index.shardCount,
          sizes: this.index.sizes,
        });
        await fs.writeFile(tmp, payload);
        await fs.rename(tmp, this.indexPath());
      });
    return this.persistChain;
  }

  async store(
    fileId: string,
    shardIndex: number,
    data: Uint8Array,
    expectedHash: string,
  ) {
    const actualHash = createHash('sha256').update(data).digest('hex');
    if (actualHash !== expectedHash) {
      throw new Error(
        `Shard hash mismatch: got ${actualHash.slice(0, 16)}..., expected ${expectedHash.slice(0, 16)}...`,
      );
    }

    const key = this.key(fileId, shardIndex);
    const incomingSize = data.length;
    const existingSize = this.index.sizes[key];
    // Only the net delta counts toward capacity when overwriting.
    const delta =
      existingSize === undefined ? incomingSize : incomingSize - existingSize;

    if (this.index.usedBytes + delta > this.capacityBytes) {
      throw new Error(
        `capacity exceeded: ${this.index.usedBytes + delta} > ${this.capacityBytes}`,
      );
    }

    const path = this.shardPath(fileId, shardIndex);
    await fs.mkdir(join(this.shardsDir(), fileId.slice(0, 2)), {
      recursive: true,
    });
    await fs.writeFile(path, data);

    if (existingSize === undefined) {
      this.index.shardCount++;
    }
    this.index.usedBytes += delta;
    this.index.sizes[key] = incomingSize;
    await this.persistIndex();
  }

  async retrieve(
    fileId: string,
    shardIndex: number,
  ): Promise<Uint8Array | null> {
    try {
      const buf = await fs.readFile(this.shardPath(fileId, shardIndex));
      return new Uint8Array(buf.buffer, buf.byteOffset, buf.byteLength);
    } catch (e: any) {
      if (e.code === 'ENOENT') return null;
      throw e;
    }
  }

  async remove(fileId: string, shardIndex: number): Promise<boolean> {
    const key = this.key(fileId, shardIndex);
    try {
      await fs.unlink(this.shardPath(fileId, shardIndex));
    } catch (e: any) {
      if (e.code === 'ENOENT') {
        // File already gone; reconcile the index if it still tracks it.
        if (this.index.sizes[key] !== undefined) {
          this.index.usedBytes -= this.index.sizes[key];
          this.index.shardCount = Math.max(0, this.index.shardCount - 1);
          delete this.index.sizes[key];
          await this.persistIndex();
        }
        return false;
      }
      throw e;
    }

    const removedSize = this.index.sizes[key];
    if (removedSize !== undefined) {
      this.index.usedBytes = Math.max(0, this.index.usedBytes - removedSize);
      this.index.shardCount = Math.max(0, this.index.shardCount - 1);
      delete this.index.sizes[key];
    }
    await this.persistIndex();
    return true;
  }

  async usedBytes(): Promise<number> {
    return this.index.usedBytes;
  }

  async shardCount(): Promise<number> {
    return this.index.shardCount;
  }

  /** Synchronous O(1) accessors backed by the in-memory index. */
  usedBytesSync(): number {
    return this.index.usedBytes;
  }

  shardCountSync(): number {
    return this.index.shardCount;
  }
}

async function walkDir(dir: string, onFile: (path: string) => Promise<void>) {
  let entries;
  try {
    entries = await fs.readdir(dir, { withFileTypes: true });
  } catch (e: any) {
    if (e.code === 'ENOENT') return;
    throw e;
  }
  for (const entry of entries) {
    const full = join(dir, entry.name);
    if (entry.isDirectory()) {
      await walkDir(full, onFile);
    } else if (entry.isFile()) {
      await onFile(full);
    }
  }
}
